flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/sink/AvroGenericRecordToRowDataMapper.java

```diff
@@ -18,44 +18,119 @@
- * time type, we can't directly use the Avro Schema converted from Iceberg schema via {@link
- * AvroSchemaUtil#convert(org.apache.iceberg.Schema, String)}.
+ * time type, we can't directly use the Avro Schema converted from Iceberg schema. Additionally,
+ * Flink's converter doesn't handle timestamp-nanos logical type properly, so we manually fix those
+ * fields after conversion.
  */
 public class AvroGenericRecordToRowDataMapper implements MapFunction<GenericRecord, RowData> {
 
   private final AvroToRowDataConverters.AvroToRowDataConverter converter;
+  private final List<Integer> timestampNanosFieldIndices;
+  private final RowType rowType;
 
-  AvroGenericRecordToRowDataMapper(RowType rowType) {
+  AvroGenericRecordToRowDataMapper(RowType rowType, List<Integer> timestampNanosFieldIndices) {
     this.converter = AvroToRowDataConverters.createRowConverter(rowType);
+    this.timestampNanosFieldIndices = timestampNanosFieldIndices;
+    this.rowType = rowType;
   }
 
   @Override
   public RowData map(GenericRecord genericRecord) throws Exception {
-    return (RowData) converter.convert(genericRecord);
+    RowData rowData = (RowData) converter.convert(genericRecord);
+
+    // Post-process: Flink's AvroToRowDataConverters doesn't properly handle timestamp-nanos,
+    // so we need to manually fix the timestamp fields after conversion
+    if (timestampNanosFieldIndices.isEmpty()) {
+      return rowData;
+    }
+
+    // Create a new GenericRowData with corrected timestamp-nanos fields
+    GenericRowData correctedRowData = new GenericRowData(rowData.getArity());
+    correctedRowData.setRowKind(rowData.getRowKind());
+
+    // Copy all fields from the converted RowData
+    for (int i = 0; i < rowData.getArity(); i++) {
+      if (timestampNanosFieldIndices.contains(i)) {
+        // Manually convert timestamp-nanos field from original Avro record
+        Object avroValue = genericRecord.get(i);
+        if (avroValue instanceof Long) {
+          long nanos = (Long) avroValue;
+          long millis = Math.floorDiv(nanos, 1_000_000);
+          int nanosOfMillis = Math.floorMod(nanos, 1_000_000);
+          TimestampData timestampData = TimestampData.fromEpochMillis(millis, nanosOfMillis);
+          correctedRowData.setField(i, timestampData);
+        } else {
+          // If not a Long, copy as-is
+          correctedRowData.setField(i, getFieldValue(rowData, i));
+        }
+      } else {
+        // Copy other fields as-is
+        correctedRowData.setField(i, getFieldValue(rowData, i));
+      }
+    }
+
+    return correctedRowData;
+  }
+
+  private Object getFieldValue(RowData rowData, int pos) {
+    if (rowData.isNullAt(pos)) {
+      return null;
+    }
+    // Use FieldGetter to extract the field value
+    LogicalType fieldType = rowType.getTypeAt(pos);
+    RowData.FieldGetter fieldGetter = RowData.createFieldGetter(fieldType, pos);
```

Comment:
Would it be possible to create the fieldgetters in the constructor when there is a nano type present?
It would save quite a bit of time if we don't recreate them every time.